hazelcast/sql.py

```diff
@@ -0,0 +1,1629 @@
+                    pass
+
+            next(iterator).add_done_callback(on_next_row)
+
+        it.add_done_callback(on_iterator_response)
+
+    When in doubt, use the blocking API shown in the first code sample.
+
+    Note that, iterators can be requested at most once per SqlResult.
+
+    One can call :func:`close` method of a result object to
+    release the resources associated with the result on the server side.
+    It might also be used to cancel query execution on the server side
+    if it is still active.
+
+    When the blocking API is used, one might also use ``with``
+    statement to automatically close the query even if an exception
+    is thrown in the iteration. ::
+
+        with client.sql.execute("SELECT ...") as result:
+            for row in result:
+                # Process the row.
+                print(row)
+
+
+    To get the number of rows updated by the query, use the
+    :func:`update_count`. ::
+
+        update_count = client.sql.execute("SELECT ...").update_count().result()
+
+    One does not have to call :func:`close` in this case, because the result
+    will already be closed in the server-side.
+    """
+
+    def __init__(self, sql_service, connection, query_id, cursor_buffer_size, execute_future):
+        self._sql_service = sql_service
+        """_InternalSqlService: Reference to the SQL service."""
+
+        self._connection = connection
+        """hazelcast.connection.Connection: Reference to the connection
+        that the execute request is made to."""
+
+        self._query_id = query_id
+        """_SqlQueryId: Unique id of the SQL query."""
+
+        self._cursor_buffer_size = cursor_buffer_size
+        """int: Size of the cursor buffer measured in the number of rows."""
+
+        self._lock = RLock()
+        """RLock: Protects the shared access to instance variables below."""
+
+        self._execute_response = Future()
+        """Future: Will be resolved with :class:`_ExecuteResponse` once the
+        execute request is resolved."""
+
+        self._iterator_requested = False
+        """bool: Flag that shows whether an iterator is already requested."""
+
+        self._closed = False
+        """bool: Flag that shows whether the query execution is still active
+        on the server side. When ``True``, there is no need to send the "close" 
+        request to the server."""
+
+        self._fetch_future = None
+        """Future: Will be set, if there are more pages to fetch on the server
+        side. It should be set to ``None`` once the fetch is completed."""
+
+        execute_future.add_done_callback(self._handle_execute_response)
+
+    def iterator(self):
+        """Returns the iterator over the result rows.
+
+        The iterator may be requested only once.
+
+        Returns:
+            Future[Iterator[Future[SqlRow]]]: Iterator that produces Future
+            of :class:`SqlRow` s. See the class documentation for the correct
+            way to use this.
```

Comment:
Raises:

HazelcastSqlError: In case of an SQL execution error.
ValueError: If the result only contains update count, or the iterator is already requested.